File: frontend/src/api/wiki.ts
import { api } from "./client";

export interface WikiPost {
  id: string;
  tenant_id: string;
  created_by_user_id: string;
  sub_org_id: string | null;
  creator_name: string;
  title: string;
  content: string;
  published: boolean;
  files: PostFile[];
  created_at: string;
  updated_at: string;
}

export interface PostFile {
  id: string;
  post_id: string;
  file_name: string;
  file_size: number;
  created_at: string;
}

export interface FSOInfo {
  name: string;
  email: string;
}

export async function listPublishedPosts(): Promise<WikiPost[]> {
  const res = await api<{ posts: WikiPost[] }>("/api/v1/wiki/");
  return res.posts;
}

export async function getFSOInfo(): Promise<FSOInfo | null> {
  const res = await api<{ fso: FSOInfo | null }>("/api/v1/wiki/fso");
  return res.fso;
}

export async function adminListPosts(subOrgId?: string): Promise<WikiPost[]> {
  const q = subOrgId ? `?sub_org_id=${encodeURIComponent(subOrgId)}` : "";
  const res = await api<{ posts: WikiPost[] }>(`/api/v1/admin/wiki/${q}`);
  return res.posts;
}

export async function adminGetPost(id: string): Promise<WikiPost> {
  return api<WikiPost>(`/api/v1/admin/wiki/${id}`);
}

export interface PostPayload {
  title: string;
  content: string;
  published: boolean;
  // Honored only for administrators by the backend. null = tenant-wide.
  sub_org_id?: string | null;
}

export async function adminCreatePost(payload: PostPayload): Promise<WikiPost> {
  return api<WikiPost>("/api/v1/admin/wiki/", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function adminUpdatePost(id: string, payload: PostPayload): Promise<WikiPost> {
  return api<WikiPost>(`/api/v1/admin/wiki/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export async function adminDeletePost(id: string): Promise<void> {
  await api(`/api/v1/admin/wiki/${id}`, { method: "DELETE" });
}

export async function adminUploadFile(postId: string, file: File): Promise<PostFile> {
  const csrf = document.cookie.match(/fastfso_csrf=([^;]*)/)?.[1];
  const headers: Record<string, string> = {};
  if (csrf) headers["X-CSRF-Token"] = decodeURIComponent(csrf);

  const body = new FormData();
  body.append("file", file);

  const res = await fetch(`/api/v1/admin/wiki/${postId}/files`, {
    method: "POST",
    headers,
    body,
    credentials: "include",
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ error: "upload failed" }));
    throw new Error((err as { error: string }).error || "upload failed");
  }
  return res.json() as Promise<PostFile>;
}

export async function adminDeleteFile(fileId: string): Promise<void> {
  await api(`/api/v1/admin/wiki/files/${fileId}`, { method: "DELETE" });
}

// Returns the URL that serves/redirects to the PDF (works for both IC and admin).
export function wikiFileViewUrl(fileId: string): string {
  return `/api/v1/wiki/files/${fileId}`;
}

File: backend/internal/wiki/wiki.go
package wiki

import (
	"time"

	"github.com/google/uuid"
)

// Post is a single wiki article belonging to a tenant.
type Post struct {
	ID              uuid.UUID  `json:"id"`
	TenantID        uuid.UUID  `json:"tenant_id"`
	CreatedByUserID uuid.UUID  `json:"created_by_user_id"`
	SubOrgID        *uuid.UUID `json:"sub_org_id"`
	Title           string     `json:"title"`
	Content         string     `json:"content"`
	Published       bool       `json:"published"`
	CreatorName     string     `json:"creator_name"`
	Files           []PostFile `json:"files"`
	CreatedAt       time.Time  `json:"created_at"`
	UpdatedAt       time.Time  `json:"updated_at"`
}

// CreateParams holds the fields needed to create a new wiki post.
type CreateParams struct {
	TenantID        uuid.UUID
	CreatedByUserID uuid.UUID
	SubOrgID        *uuid.UUID
	Title           string
	Content         string
	Published       bool
}

// UpdateParams holds the fields that can be updated on an existing wiki post.
// SubOrgID is optional: it is applied to the row only when UpdateSubOrgID is
// true, allowing callers (FSOs) to leave the sub-org untouched without having
// to re-fetch the existing value.
type UpdateParams struct {
	TenantID       uuid.UUID
	ID             uuid.UUID
	Title          string
	Content        string
	Published      bool
	SubOrgID       *uuid.UUID
	UpdateSubOrgID bool
}

// PostFile represents a PDF attached to a wiki post.
type PostFile struct {
	ID        uuid.UUID `json:"id"`
	PostID    uuid.UUID `json:"post_id"`
	FileName  string    `json:"file_name"`
	FileSize  int64     `json:"file_size"`
	CreatedAt time.Time `json:"created_at"`
}

// FSOInfo holds the name and email of an FSO for the "Your FSO" tile.
type FSOInfo struct {
	Name  string `json:"name"`
	Email string `json:"email"`
}
